### PR TITLE
[FLINK-29814][statefun] Change supported Flink version to 1.15.2

### PR DESCRIPTION
**Pull request by galenwarren**
Upgrade supported Flink version to 1.15.2.

**Note**: The e2e tests are failing, with this error:

```
4:37:02,233 ERROR org.apache.flink.statefun.e2e.smoke.SmokeRunner               - Exception in thread "main" java.lang.ClassCastException: java.lang.String cannot be cast to java.util.List
14:37:02,233 ERROR org.apache.flink.statefun.e2e.smoke.SmokeRunner               -      at org.apache.flink.statefun.flink.launcher.StatefulFunctionsClusterEntryPoint.addStatefulFunctionsConfiguration(StatefulFunctionsClusterEntryPoint.java:134)
14:37:02,233 ERROR org.apache.flink.statefun.e2e.smoke.SmokeRunner               -      at org.apache.flink.statefun.flink.launcher.StatefulFunctionsClusterEntryPoint.main(StatefulFunctionsClusterEntryPoint.java:88)
```
That line is reading `CoreOptions.ALWAYS_PARENT_FIRST_LOADER_PATTERNS_ADDITIONAL`, which was changed from a `String` to a `List<String>` in 1.15. The e2e tests depend on a published Statefun Docker image, the latest version of which is for Statefun 3.2 and which references Flink 1.14.x. So, I think a new Statefun image would need to be published to allow the e2e tests to work. I'm not sure how to do that.

**EDIT**: I think I see how to build the image locally at least, I'll try that out.

### CODE REVIEW

**Comment by galenwarren** [comment]
Thanks! Yes, you're right, and I didn't figure that out until after the first commit. I've updated the Flink base image tag and the e2e tests ran successfully locally. They're running now in CI.

**Comment by tzulitai** [comment]
@galenwarren could you try that out and report back? If all is good, then +1 to merge this.

File: README.md [hunk 1]
```diff
@@ -155,7 +155,7 @@ This section contains information for building this project.
 
 2. Build Stateful Functions Docker image: This step requires that you've already compiled artifacts from the source code.
   ```
-  $ ./tools/docker/build-distribution.sh
+  $ ./tools/docker/build-stateful-functions.sh
```
Comment (tzulitai) [inline README.md:158]: was this done by accident?

File: pom.xml [hunk 2]
```diff
@@ -79,11 +79,12 @@ under the License.
...
         <unixsocket.version>2.3.2</unixsocket.version>
         <protoc-jar-maven-plugin.version>3.11.1</protoc-jar-maven-plugin.version>
-        <flink.version>1.14.3</flink.version>
+        <flink.version>1.15.2</flink.version>
         <scala.binary.version>2.12</scala.binary.version>
         <scala.version>2.12.7</scala.version>
         <lz4-java.version>1.8.0</lz4-java.version>
-        <flink-shaded-jackson.version>2.12.4-14.0</flink-shaded-jackson.version>
+        <flink-shaded-jackson.version>2.12.4-15.0</flink-shaded-jackson.version>
+        <slf4j-log4j12.version>1.7.32</slf4j-log4j12.version>
```
Comment (tzulitai) [inline pom.xml:87]: Why is this needed now?

**Comment by galenwarren** [comment]
@tzulitai Yes, I will run the playground examples and report back.

File: statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/StatefulFunctionsConfigValidator.java [hunk 3]
```diff
@@ -57,10 +57,12 @@ private static void validateParentFirstClassloaderPatterns(Configuration configu
...
 
   private static Set<String> parentFirstClassloaderPatterns(Configuration configuration) {
-    final String[] split =
-        configuration.get(CoreOptions.ALWAYS_PARENT_FIRST_LOADER_PATTERNS_ADDITIONAL).split(";");
-    final Set<String> parentFirstClassloaderPatterns = new HashSet<>(split.length);
-    for (String s : split) {
+    final String[] patterns =
+        configuration
+            .get(CoreOptions.ALWAYS_PARENT_FIRST_LOADER_PATTERNS_ADDITIONAL)
+            .toArray(new String[0]);
```
Comment (tzulitai) [inline statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/StatefulFunctionsConfigValidator.java:63]: Probably don't really need `patterns` to be a `String[]`. Fine to just let it be a `List<String>` to keep things simple.

File: README.md (hunk 1 above)
Comment (galenwarren) [reply README.md:158]: That's actually the name of the script: `./tools/docker/build-stateful-functions.sh`. The README in the docker folder has the right name but the top-level README had `./tools/docker/build-distribution.sh`.

File: pom.xml (hunk 2 above)
Comment (galenwarren) [reply pom.xml:87]: `org.slf4j:slf4j-log4j12` is referenced in two subprojects: `statefun-smoke-e2e-common` and `statefun-flink-distribution`, and I needed to increase the version to resolve a dependency convergence issue. So I thought I'd pull the version out to the parent pom. But I'm happy to do something else if you'd prefer.

**Comment by galenwarren** [comment]
@tzulitai I'm having a couple issues with the playground tests.

First issue, which I've worked around (I think!) is that the build of the playground projects occurs in a Docker container, so it doesn't have access to my local Maven repository. I think I can work around this by building the project outside of a container and temporarily modifying the Dockerfile to skip the build step and to copy the jar file in from the host.

However, the second issue I'm running into is that I think I need to have built `apache/flink-statefun-playground:3.3.0-1.0` somewhere along the way, so that it can be accessed when `docker-compose up` is called. Is that built somewhere during the flink-statefun build process?

File: statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/StatefulFunctionsConfigValidator.java (hunk 3 above)
Comment (galenwarren) [reply statefun-flink/statefun-flink-core/src/main/java/org/apache/flink/statefun/flink/core/StatefulFunctionsConfigValidator.java]: Fixed in https://github.com/apache/flink-statefun/pull/319/commits/d2508875511442734ac9296a96354d6f6a0c965a.

**Comment by tzulitai** [comment]
@galenwarren right, its actually a bit more complicated then what I described indeed 😅 

You'll have to edit the Dockerfile to manually copy the StateFun Java SDK jar in, instead of letting it pull from Maven.

> However, the second issue I'm running into is that I think I need to have built apache/flink-statefun-playground:3.3.0-1.0

Ah, the use of the `apache/flink-statefun-playground` image for the playground examples is a more recent development I'm not familiar with. I think @tillrohrmann / @igalshilman would be able to provide info here?

**Comment by galenwarren** [comment]
OK, I got it working. I sent five messages and got what look like the expected replies:

```
(base) galen@galen-ubuntu-20:~/projects/git/coachclientconnect$ curl -X PUT -H "Content-Type: application/vnd.greeter.types/UserLogin" -d '{"user_id": "1", "user_name": "Joe", "login_type": "WEB"}' localhost:8090/greeter.fns/user/1
(base) galen@galen-ubuntu-20:~/projects/git/coachclientconnect$ curl -X GET localhost:8091/greetings
Welcome Joe!(base) galen@galen-ubuntu-20:~/projects/git/coachclientconnect$ curl -X PUT -H "Content-Type: application/vnd.greeter.types/UserLogin" -d '{"user_id": "1", "user_name": "Joe", "login_type": "WEB"}' localhost:8090/greeter.fns/user/1
(base) galen@galen-ubuntu-20:~/projects/git/coachclientconnect$ curl -X GET localhost:8091/greetings
Nice to see you again Joe.(base) galen@galen-ubuntu-20:~/projects/git/coachclientconnect$ curl -X PUT -H "Content-Type: application/vnd.greeter.types/UserLogin" -d '{"user_id": "1", "user_name": "Joe", "login_type": "WEB"}' localhost:8090/greeter.fns/user/1
(base) galen@galen-ubuntu-20:~/projects/git/coachclientconnect$ curl -X GET localhost:8091/greetings
Third time is a charm Joe!(base) galen@galen-ubuntu-20:~/projects/git/coachclientconnect$ curl -X PUT -H "Content-Type: application/vnd.greeter.types/UserLogin" -d '{"user_id": "1", "user_name": "Joe", "login_type": "WEB"}' localhost:8090/greeter.fns/user/1
(base) galen@galen-ubuntu-20:~/projects/git/coachclientconnect$ curl -X GET localhost:8091/greetings
Nice to see you for the 4th time, Joe! It has been 93593 milliseconds since we last saw you.(base) galen@galen-ubuntu-20:~/projects/git/coachclientconnect$ curl -X PUT -H "Content-Type: application/vnd.greeter.types/UserLogin" -d '{"user_id": "1", "user_name": "Joe", "login_type": "WEB"}' localhost:8090/greeter.fns/user/1
(base) galen@galen-ubuntu-20:~/projects/git/coachclientconnect$ curl -X GET localhost:8091/greetings
Nice to see you for the 5th time, Joe! It has been 7917 milliseconds since we last saw you.(base) galen@galen-ubuntu-20:~/projects/git/coachclientconnect$ [<35;39;50M[<35;24;40M[<35;13;35M```

**Review by FilKarnicki — COMMENTED** [review]
I built this project, used it in an uber jar on a standalone flink 1.15.2 cluster, ran our actual suite of tests at work and everything looks great. No issues whatsoever. 

Thank you, Warren

**Comment by tzulitai** [comment]
Thanks @galenwarren for the PR and @FilKarnicki for testing this as well.

+1 to merge, merging ...

**Comment by tzulitai** [comment]
Merged to `master`: b4a85ef031e539e7c2580a796246e2896d533d8f
